### PR TITLE
*: Bump k8s.io and controller-runtime dependencies

### PR DESCRIPTION
**Pull request by timflannagan**
# Description

This PR bumps several critical Go direct dependencies including k8s.io, go-control-plane, and controller-runtime.

## Code changes

Reverts https://github.com/solo-io/gloo/issues/7920 in the kube-based upstream plugin. Previously, we were typecasting our own lister abstraction that sits on top of client-go into concrete corev1 listers. That approach was breaking our regression suite -- likely because upstream refactored lister generation to be generic in 1.31.x -- and required us to revert that PR in order to get the regression suite back online. From my perspective, this type of typecasting violates our own abstraction layer and any performance-related issues should've been tackled closer to the source-of-truth (e.g. solo-kit).

Additionally, updates to pkg/bootstrap/leaderelector/kube/metrics.go were required after bumping controller-runtime to 0.18.x / 0.19.x to get builds back online.

## CI changes

- Update our k8s N-3 matrix from 1.29 to 1.31. Similarly, update our minimum k8s version from 1.25 to 1.27.
- Update any unit tests that were comparing proto structs via `To(Equal(...)` type operations. This is needed due to bumping the protobuf dependency as the default, underlying sizeCache value was changed. Further investigation into why this is needed will be tackled as a follow-up

# Context

This is blocking several streams of downstream work including RFE and GW API integration initiatives.

## Notes for reviewers

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
Related to https://github.com/solo-io/gloo/issues/9683

### CODE REVIEW

File: .github/workflows/.env/nightly-tests/max_versions.env [hunk 1]
```diff
@@ -1,6 +1,6 @@
-node_version='v1.29.2@sha256:51a1434a5397193442f0be2a297b488b6c919ce8a3931be0ce822606ea5ca245'
-kubectl_version='v1.29.2'
-kind_version='v0.20.0'
+node_version='v1.31.0@sha256:53df588e04085fd41ae12de0c3fe4c72f7013bba32a20e7325357a1ac94ba865'
```
Comment (timflannagan) [inline .github/workflows/.env/nightly-tests/max_versions.env:1]: Quick note: go.mod specifies the 1.31.1 patch version, but I didn't see a 1.31.1 sha image in the kind releases, so I left this here. I don't think it matters too much w.rt. patch version skew between k8s server and client versions.

File: changelog/v1.18.0-beta23/9683-k8s-1.31.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+changelog:
```
Comment (timflannagan) [inline changelog/v1.18.0-beta23/9683-k8s-1.31.yaml:1]: TODO: Fix this changelog.

File: pkg/bootstrap/leaderelector/kube/metrics.go [hunk 3]
```diff
@@ -35,3 +35,7 @@ func (s *switchAdapter) On(name string) {
 func (s *switchAdapter) Off(name string) {
 	s.gauge.WithLabelValues(name).Set(0.0)
 }
+
+func (s *switchAdapter) SlowpathExercised(name string) {
```
Comment (timflannagan) [inline pkg/bootstrap/leaderelector/kube/metrics.go:39]: Needed for controller-runtime 0.18.x due to client-go leaderelection changes.

File: projects/gateway/pkg/api/v1/kube/hack/update-codegen.sh [hunk 4]
```diff
@@ -10,10 +10,9 @@ ROOT_PKG=github.com/solo-io/gloo/projects/gateway/pkg/api/v1
 CLIENT_PKG=${ROOT_PKG}/kube/client
 APIS_PKG=${ROOT_PKG}/kube/apis
 
-# Below code is copied from https://github.com/weaveworks/flagger/blob/master/hack/update-codegen.sh
```
Comment (timflannagan) [inline projects/gateway/pkg/api/v1/kube/hack/update-codegen.sh:13]: This file is generated by solo-kit. We updated the Go template file in https://github.com/solo-io/solo-kit/pull/560. Note, this file is technically bugged and I have an open PR for fixing this in https://github.com/solo-io/solo-kit/pull/564.

File: projects/gateway/pkg/validation/robust_client_test.go [hunk 5]
```diff
@@ -65,7 +66,7 @@ var _ = Describe("RetryOnUnavailableClientConstructor", func() {
 		// sanity check
 		resp, err := client.Validate(rootCtx, &validation.GlooValidationServiceRequest{})
 		Expect(err).NotTo(HaveOccurred())
-		Expect(resp).To(Equal(res))
+		Expect(resp).To(matchers.MatchProto(res))
```
Comment (timflannagan) [inline projects/gateway/pkg/validation/robust_client_test.go:69]: Needed due to the protobuf jump that was clumped with these dependency bumps.

File: projects/gateway2/controller/controller_suite_test.go [hunk 6]
```diff
@@ -97,6 +99,13 @@ var _ = BeforeSuite(func() {
 			Port:    webhookInstallOptions.LocalServingPort,
 			CertDir: webhookInstallOptions.LocalServingCertDir,
 		}),
+		Controller: config.Controller{
+			// see https://github.com/kubernetes-sigs/controller-runtime/issues/2937
+			// in short, our tests reuse the same name (reasonably so) and the controller-runtime
+			// package does not reset the stack of controller names between tests, so we disable
+			// the name validation here.
+			SkipNameValidation: ptr.To(true),
+		},
```
Comment (timflannagan) [inline projects/gateway2/controller/controller_suite_test.go:108]: The comment calls out why this is needed, but this due to a recent c-r change that enforces stricter validation for controller names.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/solo-kit.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "enterprise.gloo.solo.io",
   "version": "v1",
-  "gen_kube_types": true,
+  "gen_kube_types": false,
```
Comment (timflannagan) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/solo-kit.json:4]: Moved this to https://github.com/solo-io/gloo/pull/10079. For context, solo-kit was generating hack/*-codegen.sh bash scripts for these nested directories that were relevant, so toggling this off / removing this option helps us manage maintenance.

File: projects/gloo/pkg/api/v1/enterprise/options/rbac/rbac.pb.uniquehash.go
Comment (timflannagan) [inline projects/gloo/pkg/api/v1/enterprise/options/rbac/rbac.pb.uniquehash.go:1]: Just a quick note on these new uniquehash.go files. This is needed as we had a bug in GME's AccessPolicy caching that required us to introduce a new primitive in this library. See https://github.com/solo-io/gloo-mesh-enterprise/pull/17392 for more information.

File: projects/gloo/pkg/api/v1/kube/client/clientset/versioned/typed/gloo.solo.io/v1/upstream.go
Comment (timflannagan) [inline projects/gloo/pkg/api/v1/kube/client/clientset/versioned/typed/gloo.solo.io/v1/upstream.go:1]: The diff in these client-gen generated files is a bit confusing. Basically, client-go had a series of improvements in 1.31 to help adopt generics and cut down on the amount of generated code for consumers of this library. The gentype package defines the common Get/List/etc. interfaces now.

File: projects/gloo/pkg/api/v1/kube/client/listers/gloo.solo.io/v1/upstream.go
Comment (timflannagan) [inline projects/gloo/pkg/api/v1/kube/client/listers/gloo.solo.io/v1/upstream.go:1]: Similar to the above comment about client-go generated code, client-go refactored the listers implementation to adopt a generics-based approach. See https://github.com/kubernetes/kubernetes/pull/121574 for more information.

File: projects/gloo/pkg/plugins/kubernetes/plugin.go
Comment (timflannagan) [inline projects/gloo/pkg/plugins/kubernetes/plugin.go:1]: I wanted to highlight this change in the sea of generated code changes. The primary change here is the removal of typecasting to corev1 listers which was causing the regression suite to fail. IMO, doing this is a violation of our own lister abstraction (that manages corev1 listers under-the-hood) and any net new issues with performance regressions could be handled as a follow-up in solo-kit.

File: test/e2e/aggregate_listener_test.go [hunk 8]
```diff
@@ -51,6 +51,9 @@ var _ = Describe("Aggregate Listener", func() {
 	JustBeforeEach(func() {
 		testContext.SetRunSettings(&gloov1.Settings{
 			Gateway: &gloov1.GatewayOptions{
+				EnableGatewayController: &wrappers.BoolValue{
+					Value: true,
+				},
```
Comment (timflannagan) [inline test/e2e/aggregate_listener_test.go:56]: Need to confirm with Tyler or Sam why this change is necessary.

Comment (sam-heilbron) [reply test/e2e/aggregate_listener_test.go:56]: Oh yes, let's chat about this, I have some context and questions

Comment (sam-heilbron) [reply test/e2e/aggregate_listener_test.go:56]: We followed up in Slack. We discussed how this is due to some unknown proto changes that affect boolean values to be overridden in our tests. This impacts just this test, because we require the `EnableGatewayController` (edge gw) to be true in Settings, but since we define some other values in the same struct, the default true value is not being respected, and instead the overriding empty value is being used so it is false.

Our plan is two-fold:
1. Keep this temporary solution to merge the large code. This way this PR doesn't go out of date
2. Immediately after, investigate what proto changes could lead to this and provide an explanation and fix

cc @timflannagan

**Review by artberger — APPROVED** [review]

**Comment by sam-heilbron** [comment]
Adding the `work in progress` label to prevent auto-merging, while we confirm this can be pulled into enterprise seamlessly

**Review by sam-heilbron — APPROVED** [review]

File: .github/workflows/pr-kubernetes-tests.yaml [hunk 9]
```diff
@@ -67,12 +67,12 @@ jobs:
 
         # September 16, 2024: 21 minutes
         - cluster-name: 'cluster-three'
-          go-test-args: '-v -timeout=25m'
+          go-test-args: '-v -timeout=30m'
```
Comment (jmhbh) [inline .github/workflows/pr-kubernetes-tests.yaml:70]: just curious why the timeout bump is necessary?

File: projects/gloo/pkg/plugins/kubernetes/plugin.go [hunk 10]
```diff
@@ -4,110 +4,94 @@ import (
...
+			kube.Kube.GetServiceNamespace(),
+		)
 	}
 
-	var sn kubelisters.ServiceNamespaceLister
-	if sn, ok = lister.(kubelisters.ServiceNamespaceLister); !ok {
-		if sl, ok := lister.(kubelisters.ServiceLister); ok {
-			sn = sl.Services(kube.Kube.GetServiceNamespace())
-		}
+	// we don't expect the old implementation to be used, but that we are keeping it as a fallback just in case
```
Comment (jmhbh) [inline projects/gloo/pkg/plugins/kubernetes/plugin.go:80]: nit: comment no longer relevant

**Review by jmhbh — APPROVED** [review]
LGTM, just a small nit and a quick q. Thanks for working on this Tim!

**Comment by jmhbh** [comment]
> Adding the `work in progress` label to prevent auto-merging, while we confirm this can be pulled into enterprise seamlessly

Created this branch `k8s-1.31-bump` to test the enterprise integration.

File: .github/workflows/pr-kubernetes-tests.yaml (hunk 9 above)
Comment (timflannagan) [reply .github/workflows/pr-kubernetes-tests.yaml:70]: @sam-heilbron Any idea on why this was necessary?

File: test/e2e/aggregate_listener_test.go (hunk 8 above)
Comment (timflannagan) [reply test/e2e/aggregate_listener_test.go:56]: Added this as a TODO on the parent issue as well.

**Comment by timflannagan** [comment]
> Created this branch k8s-1.31-bump to test the enterprise integration.

Just saw this now. I opened https://github.com/solo-io/gloo/pull/10069 earlier today to work through all the potential issues.

File: projects/gloo/pkg/plugins/kubernetes/plugin.go (hunk 10 above)
Comment (timflannagan) [reply projects/gloo/pkg/plugins/kubernetes/plugin.go]: Thanks - removed it in one of the recent commits.

**Review by jmhbh — APPROVED** [review]
